examples/dmrg/symmetric_dmrg.py

```diff
@@ -0,0 +1,179 @@
+"""
+Example of backend independent DMRG calculation for the
+spin 1/2 Heisenberg Model.
+"""
+
+from typing import Type, Text
+import tensornetwork as tn
+import numpy as np
+from tensornetwork import (FiniteMPO, FiniteMPS, U1Charge, Index,
+                           BlockSparseTensor, FiniteXXZ, FiniteDMRG)
+import jax
+#enable double precision in JAX
+jax.config.update('jax_enable_x64', True)
+
+
+def blocksparse_XXZ_mpo(Jz: np.ndarray,
+                        Jxy: np.ndarray,
+                        Bz: np.ndarray,
+                        dtype: Type[np.number] = np.float64) -> FiniteMPO:
+  """
+  Prepare a symmetric MPO.
+
+  Args:
+    Jz, Jxy, Bz: Hamiltonian parameters.
+    dtype: data type.
+
+  Returns:
+    `FiniteMPO`: The mpo of the XXZ Heisenberg model with U(1) symmetry.
+  """
+  dense_mpo = FiniteXXZ(Jz, Jxy, Bz, dtype=dtype).tensors
+  ileft = Index(U1Charge(np.array([0])), False)
+  iright = ileft.flip_flow()
+  i1 = Index(U1Charge(np.array([0, -1, 1, 0, 0])), False)
+  i2 = Index(U1Charge(np.array([0, -1, 1, 0, 0])), True)
+  i3 = Index(U1Charge(np.array([0, 1])), False)
+  i4 = Index(U1Charge(np.array([0, 1])), True)
+
+  mpotensors = [BlockSparseTensor.fromdense(
+      [ileft, i2, i3, i4], dense_mpo[0])] + [
+          BlockSparseTensor.fromdense([i1, i2, i3, i4], tensor)
+          for tensor in dense_mpo[1:-1]
+      ] + [BlockSparseTensor.fromdense([i1, iright, i3, i4], dense_mpo[-1])]
+  return FiniteMPO(mpotensors, backend='symmetric')
+
+
+def blocksparse_halffilled_spin_mps(N: int,
+                                    D: int,
+                                    B: int = 5,
+                                    dtype: Type[np.number] = np.float64):
+  """
+  Prepare a U(1) symmetric spin 1/2 MPS at zero total magnetization.
+
+  Args:
+    N: Number of spins.
+    D: The bond dimension.
+    B: The number of symmetry sectors on each ancillary link.
+    dtype: The data type of the MPS.
+
+  Returns:
+    `FiniteMPS`: A U(1) symmetric spin 1/2 mps at zero total magnetization.
+  """
+  auxcharges = [U1Charge([0])] + [
+      U1Charge.random(D, n // 2, n // 2 + B) for n in range(N - 1)
+  ] + [U1Charge([N // 2])]
+  tensors = [
+      BlockSparseTensor.random([
+          Index(auxcharges[n], False),
+          Index(U1Charge([0, 1]), False),
+          Index(auxcharges[n + 1], True)
+      ],
+                               dtype=dtype) for n in range(N)
+  ]
+  return FiniteMPS(tensors, canonicalize=True, backend='symmetric')
+
+
+def initialize_spin_mps(N: int, D: int, dtype: Type[np.number], backend: Text):
+  """
+  Helper function to initialize an MPS for a given backend.
+
+  Args:
+    N: Number of spins.
+    D: The bond dimension.
+    dtype: The data type of the MPS.
+
+  Returns:
+    `FiniteMPS`: A spin 1/2 mps for the corresponding backend.
+  """
+  if backend == 'symmetric':
+    return blocksparse_halffilled_spin_mps(N=N, D=D, B=5, dtype=dtype)
```

Comment:
Is B=5 fixed by the problem?

Reply:
no, it's a hyper parameter, but not important for the actual simulation